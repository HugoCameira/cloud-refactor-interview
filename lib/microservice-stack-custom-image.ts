// https://github.com/aws/constructs/blob/10.x/API.md#constructs-construct
import {type Construct} from 'constructs'
// https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib-readme.html
import * as cdk from 'aws-cdk-lib'
// https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_ecs-readme.html
import * as ecs from 'aws-cdk-lib/aws-ecs'
// https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_ec2-readme.html
import * as ec2 from 'aws-cdk-lib/aws-ec2'
// https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_elasticloadbalancingv2-readme.html
import * as elb from 'aws-cdk-lib/aws-elasticloadbalancingv2'
import {TargetType} from 'aws-cdk-lib/aws-elasticloadbalancingv2'

interface MicroserviceProps extends cdk.StackProps {
  serviceName: string
  pathToImage: string
  //HTTP Method
  methodType: string,
  //Path after the domain-name
  path: string,
  //What is returned by the service if illegal request
  serviceResponseFaillure: string
}

export class MicroServiceCustomImageStack extends cdk.Stack {
  constructor (scope: Construct, id: string, props: MicroserviceProps) {
    super(scope, id, props)

    const vpc = this.addVpc()
    const cluster = this.addEcsCluster(vpc)
    const lb = this.addLoadBalancer(vpc)
    this.addServer(cluster, vpc, lb, props)

    new cdk.CfnOutput(this, 'lb-endpoint', {
      value: lb.loadBalancerDnsName,
      description: 'service endpoint'
    })
  }

  private addServer (cluster: ecs.Cluster, vpc: ec2.Vpc, lb: elb.ApplicationLoadBalancer, props: MicroserviceProps): void {
    const taskDefinition = new ecs.TaskDefinition(this, 'api', {
      cpu: '512',
      compatibility: ecs.Compatibility.EC2
    })

    taskDefinition.addContainer(`${props.serviceName}-server`, {
      image: ecs.ContainerImage.fromAsset(props.pathToImage),
      portMappings: [{
        containerPort: 80,
        hostPort: 80
      }],
      memoryLimitMiB: 256,
      essential: true
    })

    const service = new ecs.Ec2Service(this, 'service', {
      cluster,
      taskDefinition
    })

    const applicationTargetGroup = new elb.ApplicationTargetGroup(this, 'app-target-group', {
      vpc,
      port: 25111,
      targetType: TargetType.INSTANCE,
      protocol: elb.ApplicationProtocol.HTTP,
      healthCheck: {
        path: '/health', // Specify the health check endpoint path
        interval: cdk.Duration.seconds(30), // Health check interval
        timeout: cdk.Duration.seconds(5), // Health check timeout
        healthyThresholdCount: 2, // Number of consecutive successful health checks to consider the target healthy
        unhealthyThresholdCount: 2, // Number of consecutive failed health checks to consider the target unhealthy
      },
    })
    service.attachToApplicationTargetGroup(applicationTargetGroup)

    const applicationListener = lb.addListener('app', {
      port: 80,
      protocol: elb.ApplicationProtocol.HTTP,
      defaultAction: elb.ListenerAction.fixedResponse(404, {
        contentType: 'text/plain',
        messageBody: `${props.serviceResponseFaillure}`
      })
    })
    applicationListener.addAction(props.serviceName, {
      priority: 5,
      conditions: [
        elb.ListenerCondition.pathPatterns([`/${props.path}/*`]),
        elb.ListenerCondition.httpRequestMethods(
          [`${props.methodType}`]
        )],
      action: elb.ListenerAction.forward([applicationTargetGroup])
    })
  }

  private addLoadBalancer (vpc: ec2.Vpc): elb.ApplicationLoadBalancer {
    const lb = new elb.ApplicationLoadBalancer(this, 'LB', {
      vpc,
      internetFacing: true,
      vpcSubnets: {
        subnetType: ec2.SubnetType.PUBLIC
      }
    })
    return lb
  }

  private addEcsCluster (vpc: ec2.Vpc): ecs.Cluster {
    const cluster = new ecs.Cluster(this, 'cluster', {
      vpc,
      capacity: {
        instanceType: ec2.InstanceType.of(ec2.InstanceClass.T2, ec2.InstanceSize.MICRO),
        maxCapacity: 2,
        minCapacity: 2,
        vpcSubnets: {
          subnetType: ec2.SubnetType.PUBLIC,
          onePerAz: true
        }
      }
    })
    return cluster
  }

  private addVpc (): ec2.Vpc {
    const vpc = new ec2.Vpc(this, 'vpc', {
      maxAzs: 3,
      subnetConfiguration: [
        {
          subnetType: ec2.SubnetType.PUBLIC,
          name: 'public'
        }
      ],
      natGateways: 0 // https://www.lastweekinaws.com/blog/the-aws-managed-nat-gateway-is-unpleasant-and-not-recommended/ :)
    })
    return vpc
  }
}
