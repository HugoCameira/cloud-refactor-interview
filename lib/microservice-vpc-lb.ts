// https://github.com/aws/constructs/blob/10.x/API.md#constructs-construct
import {type Construct} from 'constructs'
// https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_elasticloadbalancingv2-readme.html
import * as elb from 'aws-cdk-lib/aws-elasticloadbalancingv2'
// https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_ec2-readme.html
import * as ec2 from 'aws-cdk-lib/aws-ec2'
// https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib-readme.html
import * as cdk from 'aws-cdk-lib'


export class MicroServiceStackVpcLb extends cdk.Stack {

    public lb: elb.ApplicationLoadBalancer
    public vpc: ec2.Vpc
    
    constructor (scope: Construct, id: string) {
      super(scope, id, {})
      
      this.vpc = this.addVpc()
      this.lb = this.addLoadBalancer(this.vpc)
    }

    private addLoadBalancer (vpc: ec2.Vpc): elb.ApplicationLoadBalancer {
        const lb = new elb.ApplicationLoadBalancer(this, 'LB', {
          vpc,
          internetFacing: true,
          vpcSubnets: {
            subnetType: ec2.SubnetType.PUBLIC
          }
        })
        return lb
      }

    private addVpc (): ec2.Vpc {
        const vpc = new ec2.Vpc(this, 'vpc', {
          maxAzs: 3,
          subnetConfiguration: [
            {
              subnetType: ec2.SubnetType.PUBLIC,
              name: 'public'
            }
          ],
          natGateways: 0 // https://www.lastweekinaws.com/blog/the-aws-managed-nat-gateway-is-unpleasant-and-not-recommended/ :)
        })
        return vpc
      }
}


