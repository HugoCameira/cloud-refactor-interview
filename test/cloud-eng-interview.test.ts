import * as cdk from 'aws-cdk-lib';
import { Match, Template} from 'aws-cdk-lib/assertions';
import * as CloudEngInterview from '../lib/microservice-stack';

// example test. To run these tests, uncomment this file along with the
// example resource in lib/cloud-eng-interview-stack.ts

describe('Initial test suit', ()=>{

  let template: Template;

  beforeAll(()=> {
    const testApp = new cdk.App({
      outdir: 'cdk.out'
    });
    const stack = new CloudEngInterview.MicroServiceStack(testApp, 'MyTestStack',{
      serviceName: 'burrito',
      methodType: 'POST',
      path: 'wrapped',
      serviceResponseSuccess: 'This burrito is awesome!',
      serviceResponseFaillure: 'Burrito not found :(',
      env:{
          region: 'eu-west-3'
      }
    });
    template = Template.fromStack(stack);
  })

  // Check if configurations are properly configured 
  test('Check path configuration', () => {
    template.hasResourceProperties('AWS::ElasticLoadBalancingV2::ListenerRule', {
      Conditions: [
        {
         Field: "path-pattern",
         PathPatternConfig: {
          Values: [
           "/wrapped/*"
          ]
         }
        },
        {
         Field: "http-request-method",
         HttpRequestMethodConfig: {
          Values: [
           "POST"
          ]
         }
        }
       ]
    });
  });

  test('Test if ECS cluster is present', ()=>{
      template.hasResource('AWS::ECS::Cluster',{})
  });

  test('Test if LB is present', ()=>{
    template.hasResource('AWS::ElasticLoadBalancingV2::LoadBalancer',{})
  });

  // Check if VPC changes with snapshot
  test('VPC snapshot', ()=>{
    const vpc = template.findResources('AWS::EC2::VPC')
    expect(vpc).toMatchSnapshot();
  });
})