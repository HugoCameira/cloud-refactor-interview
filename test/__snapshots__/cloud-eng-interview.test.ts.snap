// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Initial test suit VPC snapshot 1`] = `
Object {
  "vpcA2121C38": Object {
    "Properties": Object {
      "CidrBlock": "10.0.0.0/16",
      "EnableDnsHostnames": true,
      "EnableDnsSupport": true,
      "InstanceTenancy": "default",
      "Tags": Array [
        Object {
          "Key": "Name",
          "Value": "MyTestStack/vpc",
        },
      ],
    },
    "Type": "AWS::EC2::VPC",
  },
}
`;
